### PR TITLE
Add control workspace and support check raft engine consistency

### PR DESCRIPTION
**Pull request by yuqi1129**
Signed-off-by: yuqi1129 <yuqi4733@gmail.com>

Subtask of #144 , this PR only support check the consistency of raft-engine files

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -63,4 +63,5 @@ protobuf = { git = "https://github.com/pingcap/rust-protobuf", rev = "82b49fea7e
 protobuf-codegen = { git = "https://github.com/pingcap/rust-protobuf", rev = "82b49fea7e696fd647b5aca0a6c6ec944eab3189" }
 
 [workspace]
-members = [ "stress" ]
+members = [ "stress", "ctl"]
```
Comment (tabokie) [inline Cargo.toml:66]:
```suggestion
members = ["stress", "ctl"]
```

File: ctl/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "ctl"
+version = "0.1.0"
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
Comment (tabokie) [inline ctl/Cargo.toml:6]: rm this line.

File: ctl/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,133 @@
+use raft_engine::Error;
+use raft_engine::Result;
+use std::path::PathBuf;
+use structopt::StructOpt;
```
Comment (tabokie) [inline ctl/src/main.rs:4]: We usually group them in this order:
std::xx; newline; third_party::xx; newline; own_library::xx;

And please copy the copyright line to this file as well.

File: src/engine.rs [hunk 4]
```diff
@@ -295,6 +295,13 @@ where
         path: &std::path::Path,
         file_builder: Arc<B>,
     ) -> Result<Vec<(u64, u64)>> {
+        if !path.exists() {
```
Comment (tabokie) [inline src/engine.rs:298]: Why raise error? We create the directory if not exist here:
https://github.com/tikv/raft-engine/blob/8c3e663a4a5af34bd9867489adcb683c6383aa6a/src/file_pipe_log.rs#L442-L445

File: ctl/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+        Ok(())
+    }
+}
+
+fn main() {
+    let opt = ControlOpt::from_args();
+    let result = opt.validate_and_execute();
+
+    if let Err(e) = result {
```
Comment (tabokie) [inline ctl/src/main.rs:129]:
```suggestion
    if let Err(e) = opt.validate_and_execute() {
```

File: ctl/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    #[structopt[subcommand]]
+    cmd: Option<Cmd>,
+}
+
+#[derive(Debug, StructOpt)]
+pub enum Cmd {
+    /// dump out all operations in log files
+    Dump {
+        /// Path of a specific raft-engine file
+        #[structopt(short, long)]
```
Comment (tabokie) [inline ctl/src/main.rs:19]: maybe `required_unless = "path"`?

File: ctl/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        /// raft_group ids(optional), format: raft_groups_id1,raft_group_id2....
+        #[structopt(short, long)]
+        raft_groups: Option<String>,
+    },
+
+    /// check log files for logical errors
+    Check {
+        /// Path of directory
+        #[structopt(short, long)]
+        path: Option<String>,
```
Comment (tabokie) [inline ctl/src/main.rs:35]: This works?
```suggestion
        path: String,
```

File: ctl/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    },
+
+    /// check log files for logical errors
+    Truncate {
+        /// Path of raft-engine storage directory
+        #[structopt(short, long)]
+        path: Option<String>,
```
Comment (tabokie) [inline ctl/src/main.rs:42]: ditto

File: ctl/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+        /// raft_group ids(optional), format: raft_groups_id1,raft_group_id2....
+        #[structopt(short, long)]
+        raft_groups: Option<String>,
+    },
+
+    /// repair log entry holes by fill in empty message
+    Autofill {
+        /// Path of raft-engine storage directory
+        #[structopt(short, long)]
+        path: Option<String>,
```
Comment (tabokie) [inline ctl/src/main.rs:61]: ditto

File: ctl/src/main.rs [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+        #[structopt(short, long)]
+        file: Option<String>,
+
+        /// Path of raft-engine storage directory
+        #[structopt(short, long, required_unless = "file")]
+        path: Option<String>,
+
+        /// raft_group ids(optional), format: raft_groups_id1,raft_group_id2....
+        #[structopt(short, long)]
+        raft_groups: Option<String>,
```
Comment (tabokie) [inline ctl/src/main.rs:28]: Use `Vec` like Autofill?

File: ctl/src/main.rs [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+        #[structopt(short, long)]
+        path: Option<String>,
+
+        /// queue name
+        #[structopt(short, long)]
+        queue: Option<String>,
+
+        /// raft_group ids(optional), format: raft_groups_id1,raft_group_id2....
+        #[structopt(short, long)]
+        raft_groups: Option<Vec<u32>>,
```
Comment (tabokie) [inline ctl/src/main.rs:69]: Should be u64.

File: ctl/src/main.rs [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+        match cmd {
+            Cmd::Autofill { .. } => self.auto_fill(),
+            Cmd::Dump { .. } => self.dump(),
+            Cmd::Truncate { .. } => self.truncate(),
+            Cmd::Check { path } => self.check(path),
+        }
+    }
+
+    fn auto_fill(&self) -> Result<()> {
+        todo!("unplement!")
```
Comment (tabokie) [inline ctl/src/main.rs:89]: How about we create these interfaces for `Engine` too, and leave their implementation empty?

**Review by tabokie — COMMENTED** [review]
CI failed, you should run format and clippy at your end. The commands are listed in repo readme.

File: ctl/Cargo.toml (hunk 2 above)
Comment (yuqi1129) [reply ctl/Cargo.toml:6]: This lines seems to be added by cargo automatically 😢 , I have not noticed this changed yet

File: Cargo.toml (hunk 1 above)
Comment (yuqi1129) [reply Cargo.toml:66]: 👍

File: ctl/src/main.rs (hunk 3 above)
Comment (yuqi1129) [reply ctl/src/main.rs:4]: This import order is recommended by the `cargo format`.  

> We usually group them in this order:
std::xx; newline; third_party::xx; newline; own_library::xx;

👌 :

> And please copy the copyright line to this file as well.
👌

File: src/engine.rs (hunk 4 above)
Comment (yuqi1129) [reply src/engine.rs:298]: @tabokie 
When we run `clt check --path x` , If the path `x` does not exist and the function will automatically create a directory `x` thats contains raft engine files. I think we should explicitly throw error if the path have not exist in the check logic.

File: ctl/src/main.rs (hunk 5 above)
Comment (yuqi1129) [reply ctl/src/main.rs:129]: OK

File: ctl/src/main.rs (hunk 12 above)
Comment (yuqi1129) [reply ctl/src/main.rs:89]: Sounds goods, I will do as your advice.

File: ctl/src/main.rs (hunk 10 above)
Comment (yuqi1129) [reply ctl/src/main.rs:28]: 👌

File: src/engine.rs (hunk 4 above)
Comment (tabokie) [reply src/engine.rs:298]: Okay, you can do a check in `consistency_check`, but normal `open` behavior should be kept the same.

File: ctl/src/main.rs (hunk 3 above)
Comment (tabokie) [reply ctl/src/main.rs:4]: Yes, `format` will sort the imports if they are in the same group. That's why we explicit add newlines to create multiple groups.

File: Cargo.toml [hunk 13]
```diff
@@ -63,4 +63,5 @@ protobuf = { git = "https://github.com/pingcap/rust-protobuf", rev = "82b49fea7e
 protobuf-codegen = { git = "https://github.com/pingcap/rust-protobuf", rev = "82b49fea7e696fd647b5aca0a6c6ec944eab3189" }
 
 [workspace]
-members = [ "stress" ]
+members = ["stress", "ctl"]
+default-members = ["ctl"]
```
Comment (tabokie) [inline Cargo.toml:67]: Probably this line is wrong. This means default use `ctl` instead of `raft_engine` crate, and the CI fails to locate tests.

Comment (yuqi1129) [reply Cargo.toml:67]: if we do no put `ctl` in default-members,  run `cargo build` in root will not compile `ctl` like `stress`, is that OK? @tabokie

Comment (tabokie) [reply Cargo.toml:67]: Yes

File: src/engine.rs [hunk 14]
```diff
@@ -290,6 +316,13 @@ where
...
+            return Err(Error::InvalidArgument(format!(
+                "raft-engine directory '{}' does not exist.",
+                path.to_str().unwrap().to_owned()
```
Comment (tabokie) [inline src/engine.rs:322]: reference is enough?
```suggestion
                path.to_str().unwrap()
```

File: Cargo.toml [hunk 15]
```diff
@@ -63,4 +63,4 @@ protobuf = { git = "https://github.com/pingcap/rust-protobuf", rev = "82b49fea7e
 protobuf-codegen = { git = "https://github.com/pingcap/rust-protobuf", rev = "82b49fea7e696fd647b5aca0a6c6ec944eab3189" }
 
 [workspace]
-members = [ "stress" ]
+members = ["stress", "ctl"]
```
Comment (tabokie) [inline Cargo.toml:66]: add a newline

**Review by tabokie — APPROVED** [review]
rest looks good

**Review by tabokie — COMMENTED** [review]
BTW, you should add ctl to coverage ignore list: https://github.com/tikv/raft-engine/blob/44dbcb0b3004e894dfd6c3b89df965223fcefa13/codecov.yml#L14

**Comment by yuqi1129** [comment]
>  BTW, you should add ctl to coverage ignore list:

Got it

File: src/engine.rs (hunk 14 above)
Comment (yuqi1129) [reply src/engine.rs:322]: 🆗

File: Cargo.toml (hunk 15 above)
Comment (yuqi1129) [reply Cargo.toml:66]: > add a newline
You mean we should add a empty line below the line members?

Comment (tabokie) [reply Cargo.toml:66]: yes, see https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

Comment (yuqi1129) [reply Cargo.toml]: 💯

**Pull request merged**